opentelemetry-api/src/opentelemetry/resources/__init__.py

```diff
@@ -11,3 +11,45 @@
 # WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 # See the License for the specific language governing permissions and
 # limitations under the License.
+from abc import ABC, abstractmethod
+from typing import Dict, Union
+
+
+class Resource(ABC):
```

Comment:
We haven't been using `ABC`/`abstractmethod` so far, but this if there's anywhere that it's a good idea, it's this package.

Reply:
I'm not the biggest fan of ABC's, but they do provide a great signal as to whether someone has implemented the required interface or not. 

I tend to use it only when I expect non-core maintainers to implement the interface. 

one caveat I see is that, as we add new APIs, implementations will need to be fixed to support the new method, even if it isn't used. But hopefully the spec itself is fairly concrete, or we simply don't annotate those methods with abstractmethod.